nipype/interfaces/base.py

```diff
@@ -760,29 +750,43 @@ class BaseInterface(Interface):
 
     This class cannot be instantiated.
 
+
+    Relevant Interface attributes
+    -----------------------------
+
+    ``input_spec`` points to the traited class for the inputs
+    ``output_spec`` points to the traited class for the outputs
+    ``_redirect_x`` should be set to ``True`` when the interface requires
+      connecting to a ``$DISPLAY`` (default is ``False``).
+    ``resource_monitor`` if ``False`` prevents resource-monitoring this
+      interface, if ``True`` monitoring will be enabled IFF the general
+      Nipype config is set on (``resource_monitor = true``).
+
+
     """
     input_spec = BaseInterfaceInputSpec
     _version = None
     _additional_metadata = []
     _redirect_x = False
     references_ = []
+    resource_monitor = True  # Enabled for this interface IFF enabled in the config
 
-    def __init__(self, from_file=None, **inputs):
+    def __init__(self, from_file=None, resource_monitor=None, **inputs):
         if not self.input_spec:
             raise Exception('No input_spec in class: %s' %
                             self.__class__.__name__)
 
         self.inputs = self.input_spec(**inputs)
-        self.estimated_memory_gb = 0.25
-        self.num_threads = 1
```

Comment:
Removing this breaks AFNI interfaces. We'll need to rework these.

https://github.com/nipy/nipype/blob/6a8318f632d3b3243cf7d69d1ae387aaa26afff8/nipype/interfaces/afni/base.py#L187

Reply:
Correct

Reply:
Ok, I've been looking at this. Regarding `num_threads` we have two sets of interfaces:
  - "ANTs-like" (ANTs, FreeSurfer, FSL, elastix): have an `inputs.num_threads` field.
  - AFNI, where `num_threads` is a class attribute.

We should unify this to only one option. And also make a decision about how information should flow between interfaces and nodes: should nodes set `num_threads` on the interface depending on their `n_procs` attribute. The alternative is that `n_procs` is read from the interface's `inputs` before running the node.

WDYT @satra, @effigies ?

Reply:
My thought:

- If I set it in the `Node`, that should override and set it in the input spec.
- If I don't specify it in the `Node`, but it is defined in the `Interface`, then the node should pass that through.

Something like the following (I haven't looked up current naming conventions and API):

```Python
class Node:

    def __init__(self, n_procs=None):
        super().__init__()
        self._n_procs = n_procs
        if hasattr(self.inputs, 'num_threads') and self._n_procs is not None:
                self.inputs.num_threads = self._n_procs

    @property
    def n_procs(self):
        if self._n_procs is not None:
            return self._n_procs
        elif hasattr(self.inputs, 'num_threads') and isdefined(self.inputs.num_threads):
            return self.inputs.num_threads
        else:
            return 1
```

Reply:
for `MapNode` would this be set for every instance of the interface? my thought would be yes, since the number of instances are dynamically created. and the worst case scenario is to run every instance in serial mode, with best case scenario controlled by the total number of procs available.

i like the code, defaults to 1, overrides with `Node`, and sets it for each instance. we need to make sure this carries over to `MapNode` and we have a sync between the interface `num_threads` and `_n_procs`.

re: the case for inputs vs class attributes, this one is fuzzy in my mind. on one hand, bad parallel implementations may provide different output, which means should rerun if number of threads change, but we don't do this right now (`hash_files = False`). i think this should eventually be a class attribute that can be overridden for any instance. or we should be strict and remove `hash_files` and make the code rerun if this number changes.

Reply:
I'll answer here, including [this Chris' comment](https://github.com/nipy/nipype/pull/2200#issuecomment-333618018) that I think it was meant to be here.

I do fundamentally agree with both, and my idea was along the lines @effigies has described (keep the info statically on `Node.n_procs`, but have a dynamic option with `Interface.inputs.num_threads`). Thanks for this much clearer description.

I also agree with Satra on the following. In our experience, `antsRegistration`'s `nohash` (or `hash_files` if talking about the specs) should be `False`. Unless you compiled with very specific and unoptimized flags, you'd get different results if you change `num_threads`. But also if you don't change it and `num_threads` > 1. That would justify an input, in addition to @effigies example.

For nipype 1.0 we can keep the `inputs.num_threads` (meaning, I'll normalize AFNI to behave as the others) and we can reconsider removing the input for nipype 2.0, when we have more experience on how this turned out. Sounds good?

Reply:
That sounds fine to me.

Reply:
sounds good to me as well.